### ISSUE TITLE
getitem/setitem putmask hooks

### ISSUE THREAD

**Issue by tdimitri** [Quansight/numpy-threading-extensions#23]
boolean and fancy index masking related routines can be sped up and are used very often.
Be nice if we could hook these routines.

**Comment by mattip** [comment]
We may have to hack the `tp_getitem`, `tp_setitem` to do this.

**Comment by teoliphant** [comment]
Yes, to replace "mask" and fancy-indexing, generically, we would likely need to replace the Python-level functions.   There might be a way to speed things up a bit by replacing functions on the "PyArray_ArrFuncs *f" member of the PyArray_Descr ("dtype") object.  There are functions there that also need to be swapped out to speed up sorting, and such.  See lines near 443 to 553 of ndarraytypes.h (though I may have an old version of the NumPy code checked-out).
